File: modules/protocol/769/src/main/java/candle/protocol/codec/UUIDCodec.java
package candle.protocol.codec;

import candle.types.ex.NotImplementedException;
import candle.types.resource.Identifier;

import java.io.*;
import java.util.UUID;

public class UUIDCodec extends TypeCodec<UUID> {
  private static TypeIdentifier<UUID> IDENTIFIER;

  public UUIDCodec() {
    if ( IDENTIFIER == null ) {
      IDENTIFIER = identifier("minecraft", "uuid");
    }
  }

  @Override
  UUID read( InputStream in, AbstractDataCodec codec ) throws
                                                       IOException,
                                                       NotImplementedException {
    throw new NotImplementedException("This functionality is not yet implemented!");
  }


  @Override
  void write( OutputStream out, UUID value, AbstractDataCodec dataCodec ) throws
                                                                          IOException {
    ByteArrayOutputStream uuidBuffer = new ByteArrayOutputStream();
    DataOutputStream dataOut = new DataOutputStream(uuidBuffer);
    dataOut.writeLong(value.getMostSignificantBits());
    dataOut.writeLong(value.getLeastSignificantBits());
    out.write(uuidBuffer.toByteArray());
  }

  @Override
  public Identifier namespacedIdentifier() {
    return IDENTIFIER;
  }
}

File: protocol/769/src/main/kotlin/net/candlemc/protocol/codec/UUIDCodec.kt
package net.candlemc.protocol.codec

import net.candlemc.types.exceptions.NotImplementedException
import net.candlemc.types.resource.Identifier
import java.io.*
import java.util.*

class UUIDCodec : TypeCodec<UUID>() {
    init {
        if (IDENTIFIER == null) {
            IDENTIFIER = identifier("minecraft", "uuid")
        }
    }

    @Throws(IOException::class, NotImplementedException::class)
    override fun read(inputStream: InputStream, dataCodec: AbstractDataCodec): UUID {
        val dataIn = DataInputStream(inputStream)
        val mostSignificantBits = dataIn.readLong()
        val leastSignificantBits = dataIn.readLong()
        if (mostSignificantBits == leastSignificantBits) {
            throw IOException("Invalid UUID data: most and least significant bits are identical")
        }
        return UUID(mostSignificantBits, leastSignificantBits)
    }


    @Throws(IOException::class)
    override fun write(outputStream: OutputStream, value: UUID, dataCodec: AbstractDataCodec) {
        val uuidBuffer = ByteArrayOutputStream()
        val dataOut = DataOutputStream(uuidBuffer)
        dataOut.writeLong(value.mostSignificantBits)
        dataOut.writeLong(value.leastSignificantBits)
        outputStream.write(uuidBuffer.toByteArray())
    }

    override fun namespacedIdentifier(): Identifier {
        return IDENTIFIER!!
    }

    companion object {
        private var IDENTIFIER: TypeIdentifier<UUID>? = null
    }
}
